File: ios/ScreenTimeGuardian/Domain/ScreenTimeRepository.swift
//
//  ScreenTimeRepository.swift
//  ScreenTime Guardian - Repository Layer
//

import Foundation
import Combine
import CoreData

// MARK: - ScreenTimeRepository
/// Local-First Repository mit 30 Tagen Historie
public actor ScreenTimeRepository {
    
    // MARK: - Properties
    private let localDataSource: LocalScreenTimeDataSource
    private let remoteDataSource: RemoteScreenTimeDataSource?
    private let syncManager: SyncManager
    
    // MARK: - Initialization
    public init(
        localDataSource: LocalScreenTimeDataSource,
        remoteDataSource: RemoteScreenTimeDataSource? = nil,
        syncManager: SyncManager = SyncManager()
    ) {
        self.localDataSource = localDataSource
        self.remoteDataSource = remoteDataSource
        self.syncManager = syncManager
    }
    
    // MARK: - Public Methods
    
    /// Speichert DailyUsage lokal und trigger Sync
    public func saveDailyUsage(_ usage: DailyUsage) async throws {
        try await localDataSource.saveDailyUsage(usage)
        
        // Sync zu Firebase (nur aggregierte Daten)
        if let remote = remoteDataSource {
            let syncData = usage.toSyncData()
            try? await remote.syncDailyStats(syncData)
        }
    }
    
    /// Holt Historie für die letzten X Tage
    public func getUsageHistory(days: Int = 30) async throws -> [DailyUsage] {
        let calendar = Calendar.current
        let endDate = Date()
        guard let startDate = calendar.date(byAdding: .day, value: -days, to: endDate) else {
            return []
        }
        
        return try await localDataSource.fetchUsageHistory(from: startDate, to: endDate)
    }
    
    /// Berechnet Streaks basierend auf lokalen Daten
    public func calculateStreak(type: StreakType, goalMinutes: Int) async -> Streak {
        let history = try? await getUsageHistory(days: 365) // Max für longest streak
        guard let days = history, !days.isEmpty else {
            return Streak(currentStreak: 0, longestStreak: 0, lastActiveDate: nil, streakType: type)
        }
        
        let sortedDays = days.sorted { $0.date > $1.date }
        var currentStreak = 0
        var longestStreak = 0
        var tempStreak = 0
        var lastActiveDate: Date?
        
        let calendar = Calendar.current
        var previousDate: Date?
        
        for day in sortedDays {
            let metGoal = didMeetStreakGoal(day: day, type: type, goalMinutes: goalMinutes)
            
            if metGoal {
                if let prev = previousDate,
                   calendar.isDate(day.date, inSameDayAs: calendar.date(byAdding: .day, value: -1, to: prev)!) {
                    tempStreak += 1
                } else {
                    tempStreak = 1
                }
                
                if currentStreak == 0 && currentStreak < tempStreak {
                    currentStreak = tempStreak
                    lastActiveDate = day.date
                }
                
                longestStreak = max(longestStreak, tempStreak)
                previousDate = day.date
            } else {
                if currentStreak == 0 {
                    // Streak gebrochen heute
                    break
                }
                tempStreak = 0
            }
        }
        
        return Streak(
            currentStreak: currentStreak,
            longestStreak: longestStreak,
            lastActiveDate: lastActiveDate,
            streakType: type
        )
    }
    
    // MARK: - Private Helpers
    
    private func didMeetStreakGoal(day: DailyUsage, type: StreakType, goalMinutes: Int) -> Bool {
        switch type {
        case .dailyFocus:
            return day.focusTimeSeconds >= TimeInterval(goalMinutes * 60)
        case .distractionFree:
            // Implementation depends on limit adherence tracking
            return true // Placeholder
        case .earlyStart:
            return day.focusSessionsCompleted > 0 // Simplification
        }
    }
}

// MARK: - Data Source Protocols

public protocol LocalScreenTimeDataSource {
    func saveDailyUsage(_ usage: DailyUsage) async throws
    func fetchUsageHistory(from: Date, to: Date) async throws -> [DailyUsage]
    func fetchDailyUsage(for date: Date) async throws -> DailyUsage?
    func saveLimit(_ limit: LimitEntity) async throws
    func fetchActiveLimits() async throws -> [LimitEntity]
    func deleteLimit(id: String) async throws
}

public protocol RemoteScreenTimeDataSource {
    func syncDailyStats(_ stats: AggregatedDailyStats) async throws
    func fetchLeaderboard() async throws -> [LeaderboardEntry]
}

// MARK: - CoreData Implementation

public actor CoreDataScreenTimeDataSource: LocalScreenTimeDataSource {
    
    private let container: NSPersistentContainer
    
    public init(container: NSPersistentContainer) {
        self.container = container
    }
    
    public func saveDailyUsage(_ usage: DailyUsage) async throws {
        let context = container.newBackgroundContext()
        try await context.perform {
            let entity = DailyUsageEntity(context: context)
            entity.id = usage.id
            entity.date = usage.date
            entity.totalScreenTimeSeconds = usage.totalScreenTimeSeconds
            entity.unlockCount = Int32(usage.unlockCount)
            entity.focusSessionsCompleted = Int32(usage.focusSessionsCompleted)
            entity.focusTimeSeconds = usage.focusTimeSeconds
            
            try context.save()
        }
    }
    
    public func fetchUsageHistory(from: Date, to: Date) async throws -> [DailyUsage] {
        let context = container.viewContext
        let request: NSFetchRequest<DailyUsageEntity> = DailyUsageEntity.fetchRequest()
        request.predicate = NSPredicate(format: "date >= %@ AND date <= %@", from as NSDate, to as NSDate)
        request.sortDescriptors = [NSSortDescriptor(key: "date", ascending: false)]
        
        let entities = try context.fetch(request)
        return entities.map { $0.toDomainModel() }
    }
    
    public func fetchDailyUsage(for date: Date) async throws -> DailyUsage? {
        let context = container.viewContext
        let request: NSFetchRequest<DailyUsageEntity> = DailyUsageEntity.fetchRequest()
        
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: date)
        let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay)!
        
        request.predicate = NSPredicate(format: "date >= %@ AND date < %@", startOfDay as NSDate, endOfDay as NSDate)
        request.fetchLimit = 1
        
        let entities = try context.fetch(request)
        return entities.first?.toDomainModel()
    }
    
    public func saveLimit(_ limit: LimitEntity) async throws {
        // Implementation
    }
    
    public func fetchActiveLimits() async throws -> [LimitEntity] {
        // Implementation
        return []
    }
    
    public func deleteLimit(id: String) async throws {
        // Implementation
    }
}

// MARK: - Entity Extensions

extension DailyUsageEntity {
    func toDomainModel() -> DailyUsage {
        DailyUsage(
            date: date ?? Date(),
            totalScreenTimeSeconds: totalScreenTimeSeconds,
            unlockCount: Int(unlockCount),
            appUsage: [], // Parse from JSON or relation
            categoryUsage: [:],
            focusSessionsCompleted: Int(focusSessionsCompleted),
            focusTimeSeconds: focusTimeSeconds
        )
    }
}

File: android/app/src/main/java/com/screentime/guardian/domain/ScreenTimeRepository.kt
package com.screentime.guardian.domain

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.time.LocalDate
import java.time.temporal.ChronoUnit

/**
 * Local-First Repository mit 30 Tagen Historie
 */
class ScreenTimeRepository(
    private val localDataSource: LocalScreenTimeDataSource,
    private val remoteDataSource: RemoteScreenTimeDataSource? = null
) {
    /**
     * Speichert DailyUsage lokal und trigger Sync
     */
    suspend fun saveDailyUsage(usage: DailyUsage) {
        localDataSource.saveDailyUsage(usage)
        
        // Sync zu Firebase (nur aggregierte Daten)
        remoteDataSource?.syncDailyStats(usage.toSyncData())
    }
    
    /**
     * Holt Historie für die letzten X Tage
     */
    suspend fun getUsageHistory(days: Int = 30): List<DailyUsage> {
        val endDate = LocalDate.now()
        val startDate = endDate.minusDays(days.toLong())
        return localDataSource.fetchUsageHistory(startDate, endDate)
    }
    
    /**
     * Berechnet Streaks basierend auf lokalen Daten
     */
    suspend fun calculateStreak(type: StreakType, goalMinutes: Int): Streak {
        val history = try {
            getUsageHistory(365) // Max für longest streak
        } catch (e: Exception) {
            emptyList()
        }
        
        if (history.isEmpty()) {
            return Streak(currentStreak = 0, longestStreak = 0, lastActiveDate = null, streakType = type)
        }
        
        val sortedDays = history.sortedByDescending { it.date }
        var currentStreak = 0
        var longestStreak = 0
        var tempStreak = 0
        var lastActiveDate: LocalDate? = null
        
        var previousDate: LocalDate? = null
        
        for (day in sortedDays) {
            val metGoal = didMeetStreakGoal(day, type, goalMinutes)
            
            if (metGoal) {
                if (previousDate != null && ChronoUnit.DAYS.between(day.date, previousDate) == 1L) {
                    tempStreak++
                } else {
                    tempStreak = 1
                }
                
                if (currentStreak == 0) {
                    currentStreak = tempStreak
                    lastActiveDate = day.date
                }
                
                longestStreak = maxOf(longestStreak, tempStreak)
                previousDate = day.date
            } else {
                if (currentStreak == 0) {
                    // Streak gebrochen heute
                    break
                }
                tempStreak = 0
            }
        }
        
        return Streak(
            currentStreak = currentStreak,
            longestStreak = longestStreak,
            lastActiveDate = lastActiveDate,
            streakType = type
        )
    }
    
    private fun didMeetStreakGoal(day: DailyUsage, type: StreakType, goalMinutes: Int): Boolean {
        return when (type) {
            StreakType.DAILY_FOCUS -> day.focusTimeSeconds >= goalMinutes * 60
            StreakType.DISTRACTION_FREE -> true // Placeholder - depends on limit adherence
            StreakType.EARLY_START -> day.focusSessionsCompleted > 0
        }
    }
    
    // Extension für Sync
    private fun DailyUsage.toSyncData(): AggregatedDailyStats {
        return AggregatedDailyStats(
            date = date.toString(),
            totalScreenTimeMinutes = (totalScreenTimeSeconds / 60).toInt(),
            focusTimeMinutes = (focusTimeSeconds / 60).toInt(),
            unlockCount = unlockCount,
            limitAdherenceRate = 1.0f // Placeholder
        )
    }
}

// Data Source Interfaces

interface LocalScreenTimeDataSource {
    suspend fun saveDailyUsage(usage: DailyUsage)
    suspend fun fetchUsageHistory(from: LocalDate, to: LocalDate): List<DailyUsage>
    suspend fun fetchDailyUsage(forDate: LocalDate): DailyUsage?
    suspend fun saveLimit(limit: LimitEntity)
    suspend fun fetchActiveLimits(): List<LimitEntity>
    suspend fun deleteLimit(id: String)
}

interface RemoteScreenTimeDataSource {
    suspend fun syncDailyStats(stats: AggregatedDailyStats)
    suspend fun fetchLeaderboard(): List<LeaderboardEntry>
}

// Gamification Models

data class Streak(
    val currentStreak: Int,
    val longestStreak: Int,
    val lastActiveDate: LocalDate?,
    val streakType: StreakType
)

enum class StreakType {
    DAILY_FOCUS,
    DISTRACTION_FREE,
    EARLY_START
}

data class AggregatedDailyStats(
    val date: String, // YYYY-MM-DD
    val totalScreenTimeMinutes: Int,
    val focusTimeMinutes: Int,
    val unlockCount: Int,
    val limitAdherenceRate: Float
)

data class LeaderboardEntry(
    val userHash: String,
    val score: Int,
    val rank: Int,
    val trend: Trend
)

enum class Trend {
    UP, DOWN, STABLE
}
